---
title: "ccbr1040_scTCRseq_4.18.2020"
author: "Da"
date: "4/16/2020"
output: html_document
---

Single cell TCR-seq data analysis
=======================

Background:
-----------

TCR is a specific molecule on the surface of T cell, responsible for recognizing antigens presented by the major histocompatibility complex (MHC). In humans, most of TCRs consist of α and β chains. TCR β chain is generated by the random recombination of variable (V), diversity (D) and joining (J) gene segments, which generates the highly variable complementary determining region 3 (CDR3) that is critical for the specificity and affinity of antigen recognition. CDR3 polymorphisms account for TCR diversity and allow T cells to target any endogenous or exogenous antigen. T cells will be specifically activated and expanded through recognizing disease-associated antigens, which lead to a skewed distribution of TCR repertoire. Hence, the analysis of TCR repertoire can be used to reflect the immune responses for patients.

Aim:
-----------

Beta5t expressed by cTECs is critical for the development of CD8 T cells. We aim to understand how TCR repertoire is formed during thymocyte development in b5t-knockout and control mice.


The different stages of single cells from this project range from early progenitor (Double positive (DP: CD8+/CD4+) cells) which are non-selected, and then progress gradually to become fully selected, mature CD8+ T cells.  Mutation of the PSMB11 gene (aka BETA5T) which is part of the proteasome and is exclusively expressed in cTEC (cortical Thymic Epithelial cells), and is the subunit that catalyzes cleavage of peptides which are presented by the MHC molecules to cells in the immune system (double positive cells) and is important for development of CD8+ T cells.  Along with TCR signal, there’s an upregulation of CD69 (stage 2).  It’s thought that the mutation in the proteosomal protein may change the nature of cleavage of peptides that are presented and may either make TCR’s less selected or make fewer CD8+ cells.  What initial evidence indicate is that there may be similar numbers of TCR’s between WT and KO but that there are differences in TCR types (less selected?) across WT and KO animals. The question to address with his TCR-Seq analysis centers on the clonality of TCR sequences that emerge from different stages of development along with the differences in WT (heterozygous) vs KO (mutant) mice. 


Cell hashing:
----------- 

HTO001 – CD69low CCR7 low (I) Double positive in the cortex of the Thymus (most immature, not all have T cell receptors)
HTO002 – CD69hi CCR7 low (II) Double positive in the cortex of the Thymus (has received TCR stimulus)
HTO004 – CD69hi CCR7hi (IV) CD8 single positive in the medulla of the Thymus
HTO005 – CD69low CCR7hi (V) CD8 single positive in the medulla of the Thymus (ready to leave Thymus “mature”)

Stages:
-----------
 
HT01: DP (CD8+/CD4+ stage 1, unsignaled) -> HT02: DP (CD8+/CD4+ stage 2, recognition of peptide) -> HT04: Stage 3 -> HT05: Stage 4 (CD8+, fully selected, medullary)  peripheral lymphoid organs (spleen, lymph nodes)
 
Samples:
-----------
 
1MF1 – sample set one HET mouse
1MF2 – sample set one KO mouse
2MF1 – sample set two HET mouse (poor hashing, low RNA reads, high mitochondrial content), excluded
2MF2 – sample set two KO mouse (poor hashing, low RNA reads, high mitochondrial content), excluded
3MF1 – sample set three HET mouse
3MF2 – sample set three KO mouse

Batch-correction:
-----------

There seems to be clear batch effects between sample set one (1MF1, 1MF2) and sample set three (3MF1, 3MF2) based on sample name projected on TSNE map and unsupervised clustering. Since we are going to collect more samples so having a reliable way to detect and remove technical variation between different batches is important. I tested several commonly used tools for batch removal (FastMNN, Harmony, Seurat3 standard integration, Seurat3 SCTtransform integration) on the dataset and decided the Seurat3 standard integration workflow does a good job. One can see the unsupervised clustering heatmaps, the “withoutBatchCorrection” one cluster samples mostly based on batch while after batch correction, most of them cluster based on sample condition(ko or het).
 
All the heatmaps are created using batch corrected gene expression value (integrated assay) while the Dotplot, VlnPlot and FeaturePlot are created using uncorrected data(RNA assay) as recommended by Seurat, for visualization purpose. However batch information will be taken into account and be used as latent variables when differential expression analysis is set up.



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
#suppressMessages(require(SingleCellExperiment))
library(Seurat)
library(dplyr)
library(ggplot2)
library(reshape2)
library(RColorBrewer)
library(ComplexHeatmap)
library(tidyverse)
library(Matrix)
library(SingleR)
library(cowplot)
library(SingleR)
library(Routliers)
#library("URD")
#library("DoubletFinder")
#library(scRNAseq)
library(here)
library(harmony)
library(batchelor)
library(devtools)
library(SeuratWrappers)
theme_set(theme_classic())
```

#### read in Seurat object and add new metadata
```{r}
setwd("~/Desktop/active_projects/ccbr1040_scTCRseq/analysis/")
#so = readRDS("~/Desktop/active_projects/seuratObject/ccbr1040_RObjectdata_noDoublet.rds")

so@meta.data$batch = ifelse(so@meta.data$sample_name=="1MF1", "one",
                            ifelse(so@meta.data$sample_name=="1MF2", "one",
                                   ifelse(so@meta.data$sample_name=="3MF1", "three","three")))
so@meta.data$batch = as.factor(so@meta.data$batch)

so@meta.data$new_name = ifelse(so@meta.data$sample_name=="1MF1", "het",
                               ifelse(so@meta.data$sample_name=="3MF1", "het",
                                      ifelse(so@meta.data$sample_name=="1MF2", "ko", "ko")))


# so@meta.data$new_name = ifelse(so@meta.data$sample_name=="1MF1", "het_1MF1",
#                                ifelse(so@meta.data$sample_name=="3MF1", "het_3MF1",
#                                       ifelse(so@meta.data$sample_name=="1MF2", "ko_1MF2", "ko_3MF2")))

so@meta.data$new_name = paste0(so@meta.data$new_name, "_", so@meta.data$hash.ID)
so@meta.data$new_name = gsub("mHTO0", "", so@meta.data$new_name)


# my_levels <- c("het_1MF1_1", "het_3MF1_1", "ko_1MF2_1", "ko_3MF2_1",
#                "het_1MF1_2", "het_3MF1_2", "ko_1MF2_2", "ko_3MF2_2",
#                "het_1MF1_4", "het_3MF1_4", "ko_1MF2_4", "ko_3MF2_4",
#                "het_1MF1_5", "het_3MF1_5", "ko_1MF2_5", "ko_3MF2_5")

my_levels <- c("het_1", "ko_1", "het_2", "ko_2", "het_4", "ko_4", "het_5", "ko_5")


so@meta.data$new_name = factor(x = so@meta.data$new_name, levels = my_levels)

DimPlot(so, reduction = "tsne", group.by = "hash.ID")

pdf("DimPlot_umap.pdf")
DimPlot(so, reduction = "umap", group.by = "hash.ID")
dev.off()
```

```{r}
so
```


```{r}
table(so@meta.data$hash.ID)

so = readRDS("~/Desktop/ccbr1040_scTCRseq/rawData/RObjectdata.rds") 
# RObjectdata.rds downloaded from 
# https://gypsum.palantircloud.com/workspace/vector/view/ri.vector.main.workbook.611480ac-354a-403e-8284-6c080bb2c135?branch=maggie
so = subset(so, cells=which(so$hash.ID!="Doublet"&so$hash.ID!="Negative"))



table(so@meta.data$hash.ID)


DimPlot(so, reduction = "tsne", group.by = "hash.ID")
DimPlot(so, reduction = "tsne", group.by = "orig.ident")


DimPlot(so, reduction = "tsne", group.by = "orig.ident")

table(so@meta.data$sample_name)


```


```{r}

```

#### run fastMN for integrate out the batch
```{r}
so_FastMNN <- NormalizeData(so_raw)
so_FastMNN <- FindVariableFeatures(so_FastMNN)
so_FastMNN <- RunFastMNN(object.list = SplitObject(so_FastMNN, split.by = "batch"))
so_FastMNN <- RunTSNE(so_FastMNN, reduction = "mnn", dims = 1:30)
so_FastMNN <- FindNeighbors(so_FastMNN, reduction = "mnn", dims = 1:30)
so_FastMNN <- FindClusters(so_FastMNN)

DimPlot(so_FastMNN, reduction = "tsne", group.by = "sample_name")

```

#### run harmony for integrate out the batch workflow
```{r}
so = Seurat::NormalizeData(so) 
so = FindVariableFeatures(so, selection.method = "vst", nfeatures = 2000) %>% ScaleData(verbose = FALSE) 
so = RunPCA(object = so, npcs = 20)

options(repr.plot.height = 2.5, repr.plot.width = 6)
so = so %>% RunHarmony("batch", plot_convergence = TRUE)

harmony_embeddings <- Embeddings(so, 'harmony')
harmony_embeddings[1:5, 1:5]

so <- so %>% 
    RunTSNE(reduction = "harmony", dims = 1:10) %>% 
    FindNeighbors(reduction = "harmony", dims = 1:10) %>% 
    FindClusters(resolution = 0.5) %>% 
    identity()
```

#### SCTransform workflow
```{r}
so.list <- SplitObject(so, split.by = "batch")
so.list <- so.list[c("one","three")]

for (i in 1:length(so.list)) {
    so.list[[i]] <- SCTransform(so.list[[i]], verbose = FALSE)
}

options(future.globals.maxSize= 8912896000)
so.features <- SelectIntegrationFeatures(object.list = so.list, nfeatures = 3000)
so.list <- PrepSCTIntegration(object.list = so.list, anchor.features = so.features, 
    verbose = FALSE)

so.anchors <- FindIntegrationAnchors(object.list = so.list, normalization.method = "SCT", 
    anchor.features = so.features, verbose = FALSE)
so.integrated <- IntegrateData(anchorset = so.anchors, normalization.method = "SCT", 
    verbose = FALSE)
```


#### standard workflow for integration
```{r}
so.list <- SplitObject(so, split.by = "batch")


for (i in 1:length(so.list)) {
    so.list[[i]] <- NormalizeData(so.list[[i]], verbose = FALSE)
    so.list[[i]] <- FindVariableFeatures(so.list[[i]], selection.method = "vst", 
        nfeatures = 2000, verbose = FALSE)
}

reference.list <- so.list[c("one","three")]
so.anchors <- FindIntegrationAnchors(object.list = reference.list, dims = 1:30, anchor.features = rownames(so))

so.integrated <- IntegrateData(anchorset = so.anchors, dims = 1:30)

DefaultAssay(so.integrated) <- "integrated"

so.integrated <- ScaleData(so.integrated, verbose = FALSE)
so.integrated <- RunPCA(so.integrated, npcs = 30, verbose = FALSE)
so.integrated <- RunTSNE(so.integrated, reduction = "pca", dims = 1:30)

DimPlot(so.integrated, reduction = "tsne", group.by = "sample_name")
DefaultAssay(so.integrated)

saveRDS(object = so.integrated, file = "~/Desktop/active_projects/seuratObject/ccbr1040_noDoublet_SeuratStandardBatchCorrected.5.20.2020.rds")

saveRDS(object = so.integrated, file = "~/Desktop/active_projects/seuratObject/ccbr1040_noDoublet_noTCR_SeuratStandardBatchCorrected.5.31.2020.rds")


my_levels <- c("het_1", "ko_1", "het_2", "ko_2", "het_4", "ko_4", "het_5", "ko_5")


so.integrated@meta.data$new_name = factor(x = so.integrated@meta.data$new_name, levels = my_levels)

DefaultAssay(so.integrated) = "RNA"
Idents(so.integrated) = "new_name"
so_ave = AverageExpression(object = so.integrated)


table(so.integrated@meta.data$new_name)
```

#### Added TCR data to the metadata and removed doublet (hash.ID) and saved new RObjectdata_noDoublet.rds
```{r}
tcr_barcode = read.csv("~/Desktop/active_projects/ccbr1040_scTCRseq/rawData/tcr_barcodes.csv", header = T, stringsAsFactors = F)
#rownames(tcr_barcode) = tcr_barcode$Barcode

tcr_barcode$barcode = gsub(".*_","", tcr_barcode$Barcode)
tcr_barcode$orig.ident = gsub("_.*","", tcr_barcode$Barcode)
tcr_barcode$cell_barcode = ifelse(tcr_barcode$orig.ident=="1MF1", gsub("-1", "-1_1", tcr_barcode$barcode),
                           ifelse(tcr_barcode$orig.ident=="1MF2", gsub("-1", "-1_2", tcr_barcode$barcode),
                           ifelse(tcr_barcode$orig.ident=="3MF1", gsub("-1", "-1_3", tcr_barcode$barcode),
                                  gsub("-1", "-1_4", tcr_barcode$barcode))))
rownames(tcr_barcode) = tcr_barcode$cell_barcode

so = AddMetaData(object = so, metadata = tcr_barcode)

so@meta.data$TCRa = paste(so@meta.data$cell_TRAV, so@meta.data$cell_top_alpha, so@meta.data$cell_TRAJ, sep = "&")
so@meta.data$TCRb = paste(so@meta.data$cell_TRBV, so@meta.data$cell_top_beta, so@meta.data$cell_TRBJ, sep = "&")

so = subset(so, cells=which(so$hash.ID!="Doublet"&!is.na(so$ab_pair)&so$hash.ID!="Negative"))

so@meta.data$TCRab = paste0(so@meta.data$TCRa, "&", so@meta.data$TCRb)

saveRDS(object = so, file = "~/Desktop/active_projects/seuratObject/ccbr1040_scTCRseq_RObjectdata_noDoublet.rds")

het = so@meta.data[which(so@meta.data$sample_name%in%c("1MF1","3MF1")),]
ko = so@meta.data[which(so@meta.data$sample_name%in%c("1MF2","3MF2")),]

het_1MF1 = so@meta.data[which(so@meta.data$sample_name%in%c("1MF1")),]
het_3MF1 = so@meta.data[which(so@meta.data$sample_name%in%c("3MF1")),]
ko_1MF2 = so@meta.data[which(so@meta.data$sample_name%in%c("1MF2")),]
ko_3MF2 = so@meta.data[which(so@meta.data$sample_name%in%c("3MF2")),]

DimPlot(so, group.by = "het.unique", repel = TRUE, reduction = "tsne",cols = c("lightgrey", "red"))+ ggtitle("selected tcr in b5t+/-")+ theme(legend.position="top")
length(is.na(so@meta.data$orig.ident))
is.na(so@meta.data$orig.ident) %>% table()

FeaturePlot(so, features = c("mHTO01", "mHTO02", "mHTO04", "mHTO05"), min.cutoff = "q05", max.cutoff = "q95", ncol = 2, reduction = "tsne")

Idents(so) = "hash.ID"

DimPlot(so,reduction = "tsne")
FeaturePlot(so, features = c("Cd69"),reduction = "tsne")
so@meta.data 
VlnPlot(so, features = c("Cd69", "Cd4", "Cd8a", "Cd28"), pt.size = 0.01, split.by = "hash.ID" )


VlnPlot(so, features = c("Cd8a"), pt.size = 0.01, split.by = "hash.ID", group.by = "" )

so@meta.data

het = so@meta.data[which(so@meta.data$sample_name%in%c("1MF1","3MF1")),]
het = het[-which(is.na(het$cell_TRAV)),]
ko = so@meta.data[which(so@meta.data$sample_name%in%c("1MF2","3MF2")),]

genes_of_interest = read.csv("processedData/genes_of_interest.csv", header = T, stringsAsFactors = F)
TCR_components = genes_of_interest[which(genes_of_interest$type=="TCR_components"), c("gene")]

Idents(so)= "new_name"

DefaultAssay(so) = "SCT"
Idents(so) = "new_name"
so_ave = AverageExpression(object = so)

```

#### visulize a list of genes that are known to categorize DP thymocytes that are in the cortex of the thymus
```{r}
genes_of_interest = read.csv("results/processedData/genes_of_interest.csv", header = T, stringsAsFactors = F)

DefaultAssay(so_FastMNN) = "RNA"
Idents(so_FastMNN) = "new_name"
so_ave = AverageExpression(object = so_FastMNN)

VlnPlot(so, features = c("Cd8a","Cd4","Cd5","Bax","H2-Q6","Bak1"), pt.size = 0.001)+NoLegend()

genes_of_interest$in_dataset = ifelse(genes_of_interest$gene%in%rownames(so), "yes", "no")

write.csv(genes_of_interest,"results/processedData/genes_of_interest.csv", quote = F, row.names = F)

```

#### function to create gene expression visulization
```{r}
genes_of_interest = read.csv("results/processedData/genes_of_interest.csv", header = T, stringsAsFactors = F)


mySeuratPlot = function(genes){
  
  DefaultAssay(so.integrated) = "RNA"
  
  name = deparse(substitute(genes))
  p1 = VlnPlot(so.integrated, features = genes, pt.size = 0.001)+NoLegend()
  pdf(paste0("VlnPlot_", name, ".pdf"), width = 18, height = 14)
  print(p1)
  dev.off()
  
  p2 = FeaturePlot(so.integrated, features = genes, reduction = "tsne")
  pdf(paste0("FeaturePlot_", name, ".pdf"), width = 12, height = 12)
  print(p2)
  dev.off()
  
  p3 = DotPlot(so.integrated, features = genes) + RotatedAxis()
  pdf(paste0("DotPlot_", name, ".pdf"), width = 9, height = 9)
  print(p3)
  dev.off()
  
  mat = so_ave$integrated[which(rownames(so_ave$integrated)%in%genes),]
  
  p4 = 
  ComplexHeatmap::Heatmap(matrix = t(scale(t(data.matrix(mat)))), name="Z-score", km=1, 
                        col=colorRampPalette(c("darkblue","grey","darkred"))(256), 
                        row_names_gp = gpar(fontsize = ifelse(nrow(mat) <= 30, 10, 
                                                       ifelse(30<nrow(mat)&nrow(mat) <= 45, 7,
                                                       ifelse(45<nrow(mat)&nrow(mat) <= 60, 5, 4))), 
                                            fontface = "bold"),
                        column_names_gp = gpar(fontsize = 10, fontface = "bold"),
                        cluster_columns = T ,column_title_gp = gpar(fontsize = 10, fontface = "bold"), 
                        column_names_rot = 65, width = unit(8, "cm"),
                        column_title = name)
  
  pdf(paste0("Heatmap_", name, ".pdf"), width = 10, height = 10)
  print(p4)
  dev.off()
  
  p5 = RidgePlot(so.integrated, features = genes)+NoLegend()
  pdf(paste0("RidgePlot_", name, ".pdf"), width = 18, height = 14)
  print(p5)
  dev.off()  
  
  p6 = DoHeatmap(object = so.integrated, features = genes)
  pdf(paste0("DoHeatmap_", name, ".pdf"))
  print(p6)
  dev.off() 
}

TCR_components = genes_of_interest[which(genes_of_interest$type=="TCR_components"), c("gene")]
CoReceptors_CoSignalingMolecules = genes_of_interest[which(genes_of_interest$type=="CoReceptors_CoSignalingMolecules"), c("gene")]
cytokineReceptors_chemokineReceptors = genes_of_interest[which(genes_of_interest$type=="cytokineReceptors_chemokineReceptors"), c("gene")]
adhesionMolecules_guidanceMolecules = genes_of_interest[which(genes_of_interest$type=="adhesionMolecules_guidanceMolecules"), c("gene")]
otherSurfaceMolecules = genes_of_interest[which(genes_of_interest$type=="otherSurfaceMolecules"), c("gene")]
signalMolecules = genes_of_interest[which(genes_of_interest$type=="signalMolecules"), c("gene")]
transcriptionFactors = genes_of_interest[which(genes_of_interest$type=="transcriptionFactors"), c("gene")]
otherMolecules = genes_of_interest[which(genes_of_interest$type=="otherMolecules"), c("gene")]



for (i in unique(genes_of_interest$type)){
  myGenes = genes_of_interest[which(genes_of_interest$type==i), c("gene")]
}

DoHeatmap(so, features = "Cd5")

mySeuratPlot(genes = TCR_components)

mySeuratPlot(genes = CoReceptors_CoSignalingMolecules)

mySeuratPlot(genes = cytokineReceptors_chemokineReceptors)
mySeuratPlot(genes = adhesionMolecules_guidanceMolecules)
mySeuratPlot(genes = otherSurfaceMolecules)
mySeuratPlot(genes = signalMolecules)
mySeuratPlot(genes = transcriptionFactors)
mySeuratPlot(genes = otherMolecules)


DefaultAssay(so.integrated) 
Idents(so.integrated) = "new_name"
so_ave = AverageExpression(object = so.integrated)


  var_genes  = apply(so_ave$integrated, 1, var)
  select_var <- names(sort(var_genes, decreasing=TRUE))[1:1000]
  so_ave_select = so_ave$integrated[select_var,]
  
  mat = so_ave_select
  
  p3 = 
  ComplexHeatmap::Heatmap(matrix = t(scale(t(data.matrix(mat)))), name="Z-score", km=1, 
                        col=colorRampPalette(c("darkblue","grey","darkred"))(256), 
                        row_names_gp = gpar(fontsize = ifelse(nrow(mat) <= 30, 10, 
                                                       ifelse(30<nrow(mat)&nrow(mat) <= 45, 7,
                                                       ifelse(45<nrow(mat)&nrow(mat) <= 60, 5, 4))), 
                                            fontface = "bold"),
                        column_names_gp = gpar(fontsize = 10, fontface = "bold"),
                        cluster_columns = T ,column_title_gp = gpar(fontsize = 10, fontface = "bold"), 
                        column_names_rot = 65, width = unit(10, "cm"),
                        column_title = "")
  
  pdf("Heatmap_unsupervised_clustering_1000Genes.pdf")
  print(p3)
  dev.off()
  

```
```{r}
so
so.integrated
```


```{r}
annotations = c("sample_name","hash.ID")
for (i in annotations){
  plot = DimPlot(so.integrated, reduction = "tsne", group.by = i)
  pdf(paste0("DimPlot", i, ".pdf"))
  print(plot)
  dev.off()
}

DimPlot(so.integrated, reduction = "tsne", group.by = "hash.ID")

so.integrated@meta.data
```


This uses the uncorrected, log-normalized, expression data and compares a logistic regression model predicting group membership from the expression of the gene and latent variables with a null model containing only latent variables. The test is based on this paper from Lior Pachter's group: http://dx.doi.org/10.1101/258566
```{r}
DefaultAssay(so.integrated) <- "RNA"
markers <- FindAllMarkers(so.integrated, test.use = "LR", latent.vars = "batch")

```

#### function to create TCR/component frequencies by sample
```{r}
myTCR_freqTable = function(kh, tcrPart) {

myList = list()
#bb = kh[[tcrPart]]
bb = so@meta.data[[tcrPart]]

for (tcr.name in bb) {
  aa_df3 = kh[which(kh[[tcrPart]]==tcr.name),]
  if (nrow(aa_df3)==0){
  aa_df3 = data.frame(mHTO04 = 0)
  for (i in c("mHTO01","mHTO02","mHTO04","mHTO05")){
  if (!(i %in% colnames(aa_df3))){
    aa_df3[[i]] = 0
  }
}

aa_df3 = apply(aa_df3, 2, as.integer) %>% data.frame() %>% t() %>% data.frame() 
#rownames(aa_df2) = aa
aa_df3$freq = sum(aa_df3[1,])
aa_df3[[tcrPart]] = tcr.name
rownames(aa_df3) = NULL
colnames(aa_df3)[1:5] = paste0(deparse(substitute(kh)),"_" ,colnames(aa_df3)[1:5])
} else {
aa_df3 = aa_df3$hash.ID %>% table() %>% data.frame() %>% t() %>% data.frame()

colnames(aa_df3)= as.matrix(aa_df3[1,])
theNames = colnames(aa_df3)
aa_df3 = aa_df3[2,] %>% data.frame()
colnames(aa_df3) = theNames


for (i in c("mHTO01","mHTO02","mHTO04","mHTO05")){
  if (!(i %in% colnames(aa_df3))){
    aa_df3[[i]] = 0
  }
}

aa_df3 = apply(aa_df3, 2, as.numeric) %>% data.frame() %>% t() %>% data.frame()
#rownames(aa_df2) = aa
aa_df3$freq = sum(aa_df3[1,])
aa_df3[[tcrPart]] = tcr.name
rownames(aa_df3) = NULL
colnames(aa_df3)[1:5] = paste0(deparse(substitute(kh)),"_" ,colnames(aa_df3)[1:5])
  }
myList[[tcr.name]] = aa_df3
}
my_data = do.call(dplyr::bind_rows, myList)
return(my_data)
}


createFreq = function(tcrPart){
  s1 = myTCR_freqTable(kh = het_1MF1, tcrPart = tcrPart)
  s2 = myTCR_freqTable(kh = het_3MF1, tcrPart = tcrPart)
  s3 = myTCR_freqTable(kh = ko_1MF2, tcrPart = tcrPart)
  s4 = myTCR_freqTable(kh = ko_3MF2, tcrPart = tcrPart)
  s1s2 = merge(s1, s2, tcrPart = tcrPart)
  s3s4 = merge(s3, s4, tcrPart = tcrPart)
  s1s2s3s4 = merge(s1s2, s3s4, tcrPart = tcrPart)
  return(s1s2s3s4)
}



df_cell_TRAV = createFreq("cell_TRAV")
df_cell_TRAJ = createFreq("cell_TRAJ")
df_cell_TRBV = createFreq("cell_TRBV")
df_cell_TRBJ = createFreq("cell_TRBJ")

df_cell_TCRa = createFreq("TCRa")
df_cell_TCRb = createFreq("TCRb")

df_top_alpha = createFreq("cell_top_alpha")
df_top_beta = createFreq("cell_top_beta")
df_ab_pair = createFreq("ab_pair")

df_cell_TCRab = createFreq("TCRab")

write.csv(df_cell_TRAV, "TRAV.csv", quote = F, row.names = F)
write.csv(df_cell_TRAJ, "TRAJ.csv", quote = F, row.names = F)
write.csv(df_cell_TRBV, "TRBV.csv", quote = F, row.names = F)
write.csv(df_cell_TRBJ, "TRBJ.csv", quote = F, row.names = F)
write.csv(df_cell_TCRa, "TCRa.csv", quote = F, row.names = F)
write.csv(df_cell_TCRb, "TCRb.csv", quote = F, row.names = F)
write.csv(df_cell_TCRab, "TCRab.csv", quote = F, row.names = F)
write.csv(df_top_alpha, "alpha_CDR3.csv", quote = F, row.names = F)
write.csv(df_top_beta, "beta_CDR3.csv", quote = F, row.names = F)

```


```{r}
myTCR_heatmap = function(df, sample){
  df[,-1] = apply(df[,-1], 2, function(x){x/sum(x)})
  df_select = select(df, contains(sample))
  rownames(df_select) =df[,1]
  df_select$sum = rowSums(df_select)
  df_select = df_select[which(df_select$sum>0), 1:4]
  
  mat = df_select
  
  heat = 
  ComplexHeatmap::Heatmap(matrix = t(scale(t(mat))), 
                          name="Z-score", km=1, 
                        col=colorRampPalette(rev(brewer.pal(n = 7, name ="RdYlBu")))(100), 
                        row_names_gp = gpar(fontsize = ifelse(nrow(mat) <= 30, 10, 
                                                       ifelse(30<nrow(mat)&nrow(mat) <= 45, 7,
                                                       ifelse(45<nrow(mat)&nrow(mat) <= 60, 5, 4))), fontface = "bold"),
                        column_names_gp = gpar(fontsize = 10, fontface = "bold"),
                        cluster_columns = T, column_title_gp = gpar(fontsize = 13, fontface = "bold"), 
                        column_title_side = "top", column_names_rot = 45, width = unit(5, "cm")) 
  pdf(file = paste0("heatmap_", colnames(df)[1],"_" ,sample, ".pdf"))
  print(heat)
  dev.off()
  print(heat)
  
   # mat = df_select[,2:4]
  mat$name = rownames(mat)
  mat_long = melt(mat)
  
p = 
  ggplot(mat_long, aes(x = variable, y = value, group = 1)) + geom_point() + facet_wrap(vars(name))+ geom_line()+ theme(axis.text.x = element_text(color="#993333",angle=90))
  
pdf(file = paste0("geomLine_",colnames(df)[1],"_" ,sample, ".pdf"), width = 10, height = 11)
print(p)
dev.off()
}

myTCR_heatmap(df = df_cell_TRAV, sample = "mHTO01")
myTCR_heatmap(df = df_cell_TRBV, sample = "mHTO01")
myTCR_heatmap(df = df_cell_TRAJ, sample = "mHTO01")
myTCR_heatmap(df = df_cell_TRBJ, sample = "mHTO01")

myTCR_heatmap(df = df_cell_TRAV, sample = "mHTO02")
myTCR_heatmap(df = df_cell_TRBV, sample = "mHTO02")
myTCR_heatmap(df = df_cell_TRAJ, sample = "mHTO02")
myTCR_heatmap(df = df_cell_TRBJ, sample = "mHTO02")

myTCR_heatmap(df = df_cell_TRAV, sample = "mHTO04")
myTCR_heatmap(df = df_cell_TRBV, sample = "mHTO04")
myTCR_heatmap(df = df_cell_TRAJ, sample = "mHTO04")
myTCR_heatmap(df = df_cell_TRBJ, sample = "mHTO04")

myTCR_heatmap(df = df_cell_TRAV, sample = "mHTO05")
myTCR_heatmap(df = df_cell_TRBV, sample = "mHTO05")
myTCR_heatmap(df = df_cell_TRAJ, sample = "mHTO05")
myTCR_heatmap(df = df_cell_TRBJ, sample = "mHTO05")

myTCR_heatmap(df = df_cell_TRAV, sample = "het_1MF1_mHTO")
myTCR_heatmap(df = df_cell_TRAV, sample = "het_3MF1_mHTO")
myTCR_heatmap(df = df_cell_TRAV, sample = "ko_1MF2_mHTO")
myTCR_heatmap(df = df_cell_TRAV, sample = "ko_3MF2_mHTO")

myTCR_heatmap(df = df_cell_TRAJ, sample = "het_1MF1_mHTO")
myTCR_heatmap(df = df_cell_TRAJ, sample = "het_3MF1_mHTO")
myTCR_heatmap(df = df_cell_TRAJ, sample = "ko_1MF2_mHTO")
myTCR_heatmap(df = df_cell_TRAJ, sample = "ko_3MF2_mHTO")


myTCR_heatmap(df = df_cell_TRBV, sample = "het_1MF1_mHTO")
myTCR_heatmap(df = df_cell_TRBV, sample = "het_3MF1_mHTO")
myTCR_heatmap(df = df_cell_TRBV, sample = "ko_1MF2_mHTO")
myTCR_heatmap(df = df_cell_TRBV, sample = "ko_3MF2_mHTO")

myTCR_heatmap(df = df_cell_TRBJ, sample = "het_1MF1_mHTO")
myTCR_heatmap(df = df_cell_TRBJ, sample = "het_3MF1_mHTO")
myTCR_heatmap(df = df_cell_TRBJ, sample = "ko_1MF2_mHTO")
myTCR_heatmap(df = df_cell_TRBJ, sample = "ko_3MF2_mHTO")


```

```{r}

myTCR_geom_line = function(df, sample){


 df[,-1] = apply(df[,-1], 2, function(x){x/sum(x)})
  df_select = select(df, contains(sample))
  rownames(df_select) =df[,1]
  df_select$sum = rowSums(df_select)
  df_select = df_select[which(df_select$sum>0), 1:4]
  
  mat = df_select[,2:4]
  mat$name = rownames(mat)
  mat_long = melt(mat)
  
p = 
  ggplot(mat_long, aes(x = variable, y = value, group = 1)) + geom_point() + facet_wrap(vars(name))+ geom_line()+ theme(axis.text.x = element_text(color="#993333",angle=90))
  
pdf(file = paste0("geomLine_",colnames(df)[1],"_" ,sample, ".pdf"), width = 9, height = 10)
print(p)
dev.off()

# ggsave(filename = paste0("geomLine_",colnames(df)[1],"_" ,sample, ".pdf"), plot = p)

}


myTCR_geom_line(df = df_cell_TRAV, sample = "het_1MF1_mHTO")


```


```{r}
df_cell_TRAV_perc = df_cell_TRAV
df_cell_TRAV_perc[,-1] = apply(df_cell_TRAV_perc[,-1], 2, function(x){x/sum(x)})

# look at a hashtag across samples, or samples across hashtags
df_cell_TRAV_perc_mHTO02 = select(df_cell_TRAV_perc, contains("mHTO02"))
rownames(df_cell_TRAV_perc_mHTO02) =df_cell_TRAV_perc[,1]
df_cell_TRAV_perc_mHTO02$sum = rowSums(df_cell_TRAV_perc_mHTO02)
df_cell_TRAV_perc_mHTO02 = df_cell_TRAV_perc_mHTO02[which(df_cell_TRAV_perc_mHTO02$sum>0), 1:4]


mat =df_cell_TRAV_perc_mHTO02 

myHeatmap = function(mat){
  

heat = 
ComplexHeatmap::Heatmap(matrix = t(scale(t(mat))), 
                          name="Z-score", km=1, 
                        col=colorRampPalette(rev(brewer.pal(n = 7, name ="RdYlBu")))(100), 
                        row_names_gp = gpar(fontsize = ifelse(nrow(mat) <= 30, 10, 
                                                       ifelse(30<nrow(mat)&nrow(mat) <= 45, 7,
                                                       ifelse(45<nrow(mat)&nrow(mat) <= 60, 5, 4))), fontface = "bold"),
                        column_names_gp = gpar(fontsize = 10, fontface = "bold"),
                        cluster_columns = T, column_title = deparse(substitute(mat)), column_title_gp = gpar(fontsize = 13, fontface = "bold"), 
                        column_title_side = "top", column_names_rot = 45, width = unit(5, "cm")) 
  pdf(file = paste0(deparse(substitute(mat)), ".pdf"))
  print(heat)
  dev.off()
}

myHeatmap(mat = df_cell_TRAV_perc_mHTO02)

```




```{r}
df_ko_TCRa = myTCR_freqTable(kh = ko, tcrPart = "TCRa")
df_het_TCRa = myTCR_freqTable(kh = het, tcrPart = "TCRa")
df_ko.het_TCRa = merge(df_ko_TCRa, df_het_TCRa, by="TCRa")
write.csv(df_ko.het_TCRa, "processedData/TCRa_freq.csv",quote = F, row.names = F)

df_ko_TCRb = myTCR_freqTable(kh = ko, tcrPart = "TCRb")
df_het_TCRb = myTCR_freqTable(kh = het, tcrPart = "TCRb")
df_ko.het_TCRb = merge(df_ko_TCRb, df_het_TCRb, by="TCRb")
write.csv(df_ko.het_TCRb, "processedData/TCRb_freq.csv",quote = F, row.names = F)

df_ko_cell_TRAV = myTCR_freqTable(kh = ko, tcrPart = "cell_TRAV")
df_het_cell_TRAV = myTCR_freqTable(kh = het, tcrPart = "cell_TRAV")
df_ko.het_cell_TRAV = merge(df_ko_cell_TRAV, df_het_cell_TRAV, by="cell_TRAV")
write.csv(df_ko.het_cell_TRAV, "processedData/cell_TRAV_freq.csv",quote = F, row.names = F)

df_ko_cell_TRBV = myTCR_freqTable(kh = ko, tcrPart = "cell_TRBV")
df_het_cell_TRBV = myTCR_freqTable(kh = het, tcrPart = "cell_TRBV")
df_ko.het_cell_TRBV = merge(df_ko_cell_TRBV, df_het_cell_TRBV, by="cell_TRBV")
write.csv(df_ko.het_cell_TRBV, "processedData/cell_TRBV_freq.csv",quote = F, row.names = F)

df_ko_cell_TRAJ = myTCR_freqTable(kh = ko, tcrPart = "cell_TRAJ")
df_het_cell_TRAJ = myTCR_freqTable(kh = het, tcrPart = "cell_TRAJ")
df_ko.het_cell_TRAJ = merge(df_ko_cell_TRAJ, df_het_cell_TRAJ, by="cell_TRAJ")
write.csv(df_ko.het_cell_TRAJ, "processedData/cell_TRAJ_freq.csv",quote = F, row.names = F)

df_ko_cell_TRBJ = myTCR_freqTable(kh = ko, tcrPart = "cell_TRBJ")
df_het_cell_TRBJ = myTCR_freqTable(kh = het, tcrPart = "cell_TRBJ")
df_ko.het_cell_TRBJ = merge(df_ko_cell_TRBJ, df_het_cell_TRBJ, by="cell_TRBJ")
write.csv(df_ko.het_cell_TRBJ, "processedData/cell_TRBJ_freq.csv",quote = F, row.names = F)

df_ko_cell_top_alpha = myTCR_freqTable(kh = ko, tcrPart = "cell_top_alpha")
df_het_cell_top_alpha = myTCR_freqTable(kh = het, tcrPart = "cell_top_alpha")
df_ko.het_cell_top_alpha = merge(df_ko_cell_top_alpha, df_het_cell_top_alpha, by="cell_top_alpha")
write.csv(df_ko.het_cell_top_alpha, "processedData/cell_top_alpha_freq.csv",quote = F, row.names = F)

df_ko_cell_top_beta = myTCR_freqTable(kh = ko, tcrPart = "cell_top_beta")
df_het_cell_top_beta = myTCR_freqTable(kh = het, tcrPart = "cell_top_beta")
df_ko.het_cell_top_beta = merge(df_ko_cell_top_beta, df_het_cell_top_beta, by="cell_top_beta")
write.csv(df_ko.het_cell_top_beta, "processedData/cell_top_beta_freq.csv",quote = F, row.names = F)

df_ko_ab_pair = myTCR_freqTable(kh = ko, tcrPart = "ab_pair")
df_het_ab_pair = myTCR_freqTable(kh = het, tcrPart = "ab_pair")
df_ko.het_ab_pair = merge(df_ko_ab_pair, df_het_ab_pair, by="ab_pair")
write.csv(df_ko.het_ab_pair, "processedData/ab_pair_freq.csv",quote = F, row.names = F)
```




```{r}
hetf = ko_het_alpha_freq[which(ko_het_alpha_freq$ko_freq==0),]
hetf = ko_het_beta_freq[which(ko_het_beta_freq$ko_freq==0),]

hetf_select = hetf[which(hetf$het_freq>1),1]
so@meta.data$het.unique = ifelse(so@meta.data[[tcr]]%in%hetf_select, "yes","no")
hetf = melt(hetf[1:20,])

kof = ko_het_alpha_freq[which(ko_het_alpha_freq$het_freq==0),]
kof = ko_het_beta_freq[which(ko_het_beta_freq$het_freq==0),]

kof_select = kof[which(kof$ko_freq>1),1]
so@meta.data$ko.unique = ifelse(so@meta.data[[tcr]]%in%kof_select, "yes","no")
kof = melt(kof[1:20,])



ggplot(hetf, aes(x = Var1, y = value, fill=variable)) + geom_bar(stat = "identity", position = "dodge") + coord_flip()+
  ylab("Frequency") + xlab(tcr) + theme(legend.position="top")
DimPlot(so, group.by = "het.unique", repel = TRUE, reduction = "tsne",cols = c("lightgrey", "red"))+ 
ggtitle(paste("selected",tcr, "in b5t+/-"))+ theme(legend.position="top")


```


```{r}
alpha_select = subset(so, cells=which(so@meta.data$het.unique=="yes"|so@meta.data$ko.unique=="yes"))

DimPlot(alpha_select,group.by = "ko.unique", reduction = "tsne")
alpha_select@meta.data
```


```{r}
Idents(alpha_select) = "ko.unique"
ko.unique.markers =
FindMarkers(alpha_select, ident.1="yes", ident.2="no",test.use="MAST", logfc.threshold = 0.25, min.pct = 0.25)

ko.unique.markers.heatmap = DoHeatmap(alpha_select, features = rownames(ko.unique.markers))+ NoLegend()

ko.unique.aveExpr=
AverageExpression(alpha_select, features = rownames(ko.unique.markers))

mat =ko.unique.aveExpr$RNA



ComplexHeatmap::Heatmap(matrix = t(scale(t(mat))), 
                          name="Z-score", km=1, 
                        col=colorRampPalette(rev(brewer.pal(n = 7, name ="RdYlBu")))(100), 
                        row_names_gp = gpar(fontsize = ifelse(nrow(mat) <= 30, 10, 
                                                       ifelse(30<nrow(mat)&nrow(mat) <= 45, 7,
                                                       ifelse(45<nrow(mat)&nrow(mat) <= 60, 5, 4))), fontface = "bold"),
                        column_names_gp = gpar(fontsize = 10, fontface = "bold"),
                        cluster_columns = T, column_title = "gene expression", column_title_gp = gpar(fontsize = 13, fontface = "bold"), 
                        column_title_side = "top", column_names_rot = 45, width = unit(5, "cm")) 

so@meta.data

```


```{r}
het_HTO01 = het[which(het$HTO_maxID=="mHTO01"&het[[tcr]]!="NA&NA&NA"),]
het_HTO02 = het[which(het$HTO_maxID=="mHTO02"&het[[tcr]]!="NA&NA&NA"),]
het_HTO04 = het[which(het$HTO_maxID=="mHTO04"&het[[tcr]]!="NA&NA&NA"),]
het_HTO05 = het[which(het$HTO_maxID=="mHTO05"&het[[tcr]]!="NA&NA&NA"),]


ko_HTO01 = ko[which(ko$HTO_maxID=="mHTO01"&ko[[tcr]]!="NA&NA&NA"),]
ko_HTO02 = ko[which(ko$HTO_maxID=="mHTO02"&ko[[tcr]]!="NA&NA&NA"),]
ko_HTO04 = ko[which(ko$HTO_maxID=="mHTO04"&ko[[tcr]]!="NA&NA&NA"),]
ko_HTO05 = ko[which(ko$HTO_maxID=="mHTO05"&ko[[tcr]]!="NA&NA&NA"),]

write.csv(het_HTO05, "het_HTO05.csv", quote = F)

so@meta.data

Idents(so) = "HTO_maxID"
DimPlot(so, reduction = "tsne")

het_HTO01[[tcr]] %>% table() %>%plot(las=2)
het_HTO02[[tcr]] %>% table() %>%plot(las=2)
het_HTO04[[tcr]] %>% table() %>%plot(las=2)
het_HTO05[[tcr]] %>% table() %>%plot(las=2)
library(VennDiagram)
venn.plot <- venn.diagram(list(het_HTO01[[tcr]],
                                     het_HTO02[[tcr]],
                                     het_HTO04[[tcr]],
                                     het_HTO05[[tcr]]), NULL, 
                                fill=c("blue", "red","grey","green"), lwd = 1, cat.cex = 2,
                                alpha=c(0.5,0.5,0.5,0.5), cex = 3, cat.fontface=4, 
                                category.names=c("het_HTO01","het_HTO02","het_HTO04","het_HTO05"))
#Het
venn.plot <- venn.diagram(list(het_HTO02[[tcr]],
                                     het_HTO04[[tcr]],
                                     het_HTO05[[tcr]]), NULL, 
                                fill=c("blue", "red","grey"), lwd = 1, cat.cex = 2,
                                alpha=c(0.5,0.5,0.5), cex = 3, cat.fontface=4, 
                                category.names=c("het_HTO02","het_HTO04","het_HTO05"))

#ko
venn.plot <- venn.diagram(list(ko_HTO02[[tcr]],
                                     ko_HTO04[[tcr]],
                                     ko_HTO05[[tcr]]), NULL, 
                                fill=c("blue", "red","grey"), lwd = 1, cat.cex = 2,
                                alpha=c(0.5,0.5,0.5), cex = 3, cat.fontface=4, 
                                category.names=c("ko_HTO02","ko_HTO04","ko_HTO05"))

plot_grid(venn.plot)
```


```{r}
het = so@meta.data[which(so@meta.data$sample_name%in%c("1MF1","3MF1")),]
het = het[-which(is.na(het$cell_TRAV)),]
ko = so@meta.data[which(so@meta.data$sample_name%in%c("1MF2","3MF2")),]
ko = ko[-which(is.na(ko$cell_TRAV)),]

het_1MF1 = so@meta.data[which(so@meta.data$sample_name%in%c("1MF1")),]
het_3MF1 = so@meta.data[which(so@meta.data$sample_name%in%c("3MF1")),]
ko_1MF2 = so@meta.data[which(so@meta.data$sample_name%in%c("1MF2")),]
ko_3MF2 = so@meta.data[which(so@meta.data$sample_name%in%c("3MF2")),]

het_1MF1 = het[which(het$sample_name%in%c("1MF1")),]
het_3MF1 = het[which(het$sample_name%in%c("3MF1")),]
ko_1MF2 =  ko[which(ko$sample_name%in%c("1MF2")),]
ko_3MF2 =  ko[which(ko$sample_name%in%c("3MF2")),]

myFrequentTCR = function(tcr,n){
  all = c(het_1MF1[[tcr]],het_3MF1[[tcr]],ko_1MF2[[tcr]],ko_3MF2[[tcr]]) %>% table()%>%data.frame()
  all = all[order(all$Freq, decreasing = T),]
  colnames(all) = paste(tcr,colnames(all), sep = "_")
  colnames(all)[1] = str_sub(string = colnames(all)[1],start = 1,end = -3)
  all_top = all[1:n,1]
  return(all_top)
}

all_TRAV_Top = myFrequentTCR(tcr = "cell_TRAV", n = 40)
all_TRBV_Top = myFrequentTCR(tcr = "cell_TRBV", n = 22)
all_alpha_CDR3_Top = myFrequentTCR(tcr = "cell_top_alpha", n = 40) #CDR3 sequence


tcr = "cell_TRBV"
all = c(het_1MF1[[tcr]],het_3MF1[[tcr]],ko_1MF2[[tcr]],ko_3MF2[[tcr]]) %>% table()%>%data.frame()
```


```{r}
all_TRAV = c(het_1MF1$cell_TRAV,het_3MF1$cell_TRAV,ko_1MF2$cell_TRAV,ko_3MF2$cell_TRAV) %>% table()%>%data.frame()
all_TRAV = all_TRAV[order(all_TRAV$Freq, decreasing = T),]
colnames(all_TRAV) = paste("all_TRAV",colnames(all_TRAV), sep = "_")
colnames(all_TRAV)[1] = str_sub(string = colnames(all_TRAV)[1],start = 1,end = -3)
all_TRAV_Top20 = all_TRAV[1:40,1]

myTCR = function(tcr,sample, top.part){
  sample.part = sample[[tcr]] %>% table()%>%data.frame() 
  sample.part$fraction = sample.part[,2]/length(sample[[tcr]])
  colnames(sample.part) = paste(deparse(substitute(sample)),colnames(sample.part), sep = "_")
  colnames(sample.part)[1] = str_sub(string = colnames(sample.part)[1],start = 1,end = -3)
  sample.part_n = sample.part[which(sample.part[,1]%in%top.part),c(1,3)]
  colnames(sample.part_n)[1] =  tcr
  return(sample.part_n)
}

#TRAV
het_1MF1_TRAV_20 = myTCR(tcr = "cell_TRAV",sample = het_1MF1,top.part = all_TRAV_Top20)
het_3MF1_TRAV_20 = myTCR(tcr = "cell_TRAV",sample = het_3MF1,top.part = all_TRAV_Top20)
ko_1MF2_TRAV_20 = myTCR(tcr = "cell_TRAV",sample = ko_1MF2,top.part = all_TRAV_Top20)
ko_3MF2_TRAV_20 = myTCR(tcr = "cell_TRAV",sample = ko_3MF2,top.part = all_TRAV_Top20)

TRAV = merge(het_1MF1_TRAV_20, het_3MF1_TRAV_20 ,by="cell_TRAV")%>%
merge(ko_1MF2_TRAV_20,by="cell_TRAV") %>%
merge(ko_3MF2_TRAV_20 ,by="cell_TRAV")
colnames(TRAV) = gsub("_fraction","",colnames(TRAV))
rownames(TRAV) = TRAV$cell_TRAV
TRAV$cell_TRAV = NULL

#TRBV
het_1MF1_TRBV_20 = myTCR(tcr = "cell_TRBV",sample = het_1MF1,top.part = all_TRBV_Top)
het_3MF1_TRBV_20 = myTCR(tcr = "cell_TRBV",sample = het_3MF1,top.part = all_TRBV_Top)
ko_1MF2_TRBV_20 = myTCR(tcr = "cell_TRBV",sample = ko_1MF2,top.part = all_TRBV_Top)
ko_3MF2_TRBV_20 = myTCR(tcr = "cell_TRBV",sample = ko_3MF2,top.part = all_TRBV_Top)

TRBV = merge(het_1MF1_TRBV_20, het_3MF1_TRBV_20 ,by="cell_TRBV")%>%
merge(ko_1MF2_TRBV_20,by="cell_TRBV") %>%
merge(ko_3MF2_TRBV_20 ,by="cell_TRBV")
colnames(TRBV) = gsub("_fraction","",colnames(TRBV))
rownames(TRBV) = TRBV$cell_TRBV
TRBV$cell_TRBV = NULL


#alpha CDR3
het_1MF1_alpha_CDR3_Top = myTCR(tcr = "cell_top_alpha",sample = het_1MF1,top.part = all_alpha_CDR3_Top)
het_3MF1_alpha_CDR3_Top = myTCR(tcr = "cell_top_alpha",sample = het_3MF1,top.part = all_alpha_CDR3_Top)
ko_1MF2_alpha_CDR3_Top = myTCR(tcr = "cell_top_alpha",sample = ko_1MF2,top.part = all_alpha_CDR3_Top)
ko_3MF2_alpha_CDR3_Top = myTCR(tcr = "cell_top_alpha",sample = ko_3MF2,top.part = all_alpha_CDR3_Top)



alpha_CDR3 = merge(het_1MF1_alpha_CDR3_Top, het_3MF1_alpha_CDR3_Top ,by="cell_top_alpha")%>%
merge(ko_1MF2_alpha_CDR3_Top,by="cell_top_alpha") %>%
merge(ko_3MF2_alpha_CDR3_Top ,by="cell_top_alpha")
colnames(alpha_CDR3) = gsub("_fraction","",colnames(alpha_CDR3))
rownames(alpha_CDR3) = alpha_CDR3$cell_top_alpha
alpha_CDR3$cell_top_alpha = NULL



TRAV_long = melt(TRAV)
ggplot(TRAV_long, aes(x = cell_TRAV, y = value, fill=variable)) + geom_bar(stat = "identity", position = "dodge") + coord_flip()



tcr = "cell_TRAV"
sample = het_1MF1


sample.part = sample[[tcr]] %>% table()%>%data.frame() 
sample.part$fraction = sample.part[,2]/length(sample[[tcr]])


colnames(het_1MF1_TRAV) = paste("het_1MF1_TRAV",colnames(het_1MF1_TRAV), sep = "_")
colnames(het_1MF1_TRAV)[1] = str_sub(string = colnames(het_1MF1_TRAV)[1],start = 1,end = -3)
het_1MF1_TRAV_20 = het_1MF1_TRAV[which(het_1MF1_TRAV[,1]%in%all_TRAV_Top20),c(1,3)]
colnames(het_1MF1_TRAV_20)[1] = "cell_TRAV"


het_3MF1_TRAV = het_3MF1$cell_TRAV %>% table()%>%data.frame()
ko_1MF2_TRAV = ko_1MF2$cell_TRAV %>% table()%>%data.frame()
ko_3MF2_TRAV = ko_3MF2$cell_TRAV %>% table()%>%data.frame()


so@meta.data

```
