---
title: "ccbr1040_scTCRseq_4.18.2020"
author: "Da"
date: "4/16/2020"
output: html_document
---

Background:

TCR is a specific molecule on the surface of T cell, responsible for recognizing antigens presented by the major histocompatibility complex (MHC). In humans, most of TCRs consist of α and β chains. TCR β chain is generated by the random recombination of variable (V), diversity (D) and joining (J) gene segments, which generates the highly variable complementary determining region 3 (CDR3) that is critical for the specificity and affinity of antigen recognition. CDR3 polymorphisms account for TCR diversity and allow T cells to target any endogenous or exogenous antigen. T cells will be specifically activated and expanded through recognizing disease-associated antigens, which lead to a skewed distribution of TCR repertoire. Hence, the analysis of TCR repertoire can be used to reflect the immune responses for patients.

Aim:

Beta5t expressed by cTECs is critical for the development of CD8 T cells. We aim to understand how TCR repertoire is formed during thymocyte development in b5t-knockout and control mice.




```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
suppressMessages(require(SingleCellExperiment))
suppressMessages(require(Seurat))
library(dplyr)
library(ggplot2)
library(reshape2)
library(RColorBrewer)
library(ComplexHeatmap)
library(tidyverse)


suppressMessages(require(Matrix))
suppressMessages(require(SingleR))
suppressMessages(require(cowplot))
library(SingleR)
library(Routliers)
library("URD")
library("DoubletFinder")
library(scRNAseq)

theme_set(theme_classic())
```


```{r}
setwd("~/Desktop/active_projects/ccbr1040_scTCRseq/analysis/")
so = readRDS("~/Desktop/active_projects/ccbr1040_scTCRseq/rawData/RObjectdata.rds")


```

#### Added TCR data to the metadata and removed doublet (hash.ID) and saved new RObjectdata_noDoublet.rds
```{r}
tcr_barcode = read.csv("~/Desktop/active_projects/ccbr1040_scTCRseq/rawData/tcr_barcodes.csv", header = T, stringsAsFactors = F)
#rownames(tcr_barcode) = tcr_barcode$Barcode

tcr_barcode$barcode = gsub(".*_","", tcr_barcode$Barcode)
tcr_barcode$orig.ident = gsub("_.*","", tcr_barcode$Barcode)
tcr_barcode$cell_barcode = ifelse(tcr_barcode$orig.ident=="1MF1", gsub("-1", "-1_1", tcr_barcode$barcode),
                           ifelse(tcr_barcode$orig.ident=="1MF2", gsub("-1", "-1_2", tcr_barcode$barcode),
                           ifelse(tcr_barcode$orig.ident=="3MF1", gsub("-1", "-1_3", tcr_barcode$barcode),
                                  gsub("-1", "-1_4", tcr_barcode$barcode))))
rownames(tcr_barcode) = tcr_barcode$cell_barcode




so = AddMetaData(object = so, metadata = tcr_barcode)

so@meta.data$TCRa = paste(so@meta.data$cell_TRAV, so@meta.data$cell_top_alpha, so@meta.data$cell_TRAJ, sep = "&")
so@meta.data$TCRb = paste(so@meta.data$cell_TRBV, so@meta.data$cell_top_beta, so@meta.data$cell_TRBJ, sep = "&")



so = subset(so, cells=which(so$hash.ID!="Doublet"&!is.na(so$ab_pair)&so$hash.ID!="Negative"))

so@meta.data$TCRab = paste0(so@meta.data$TCRa, "&", so@meta.data$TCRb)

saveRDS(object = so, file = "~/Desktop/active_projects/seuratObject/ccbr1040_scTCRseq_RObjectdata_noDoublet.rds")

het = so@meta.data[which(so@meta.data$sample_name%in%c("1MF1","3MF1")),]
ko = so@meta.data[which(so@meta.data$sample_name%in%c("1MF2","3MF2")),]

het_1MF1 = so@meta.data[which(so@meta.data$sample_name%in%c("1MF1")),]
het_3MF1 = so@meta.data[which(so@meta.data$sample_name%in%c("3MF1")),]
ko_1MF2 = so@meta.data[which(so@meta.data$sample_name%in%c("1MF2")),]
ko_3MF2 = so@meta.data[which(so@meta.data$sample_name%in%c("3MF2")),]




DimPlot(so, group.by = "het.unique", repel = TRUE, reduction = "tsne",cols = c("lightgrey", "red"))+ ggtitle("selected tcr in b5t+/-")+ theme(legend.position="top")
length(is.na(so@meta.data$orig.ident))
is.na(so@meta.data$orig.ident) %>% table()

FeaturePlot(so, features = c("mHTO01", "mHTO02", "mHTO04", "mHTO05"), min.cutoff = "q05", max.cutoff = "q95", ncol = 2, reduction = "tsne")

Idents(so) = "hash.ID"

DimPlot(so,reduction = "tsne")


FeaturePlot(so, features = c("Cd69"),reduction = "tsne")

```



```{r}
myTCR_freqTable = function(kh, tcrPart) {

myList = list()
#bb = kh[[tcrPart]]
bb = so@meta.data[[tcrPart]]

for (tcr.name in bb) {
  aa_df3 = kh[which(kh[[tcrPart]]==tcr.name),]
  if (nrow(aa_df3)==0){
  aa_df3 = data.frame(mHTO04 = 0)
  for (i in c("mHTO01","mHTO02","mHTO04","mHTO05")){
  if (!(i %in% colnames(aa_df3))){
    aa_df3[[i]] = 0
  }
}

aa_df3 = apply(aa_df3, 2, as.integer) %>% data.frame() %>% t() %>% data.frame() 
#rownames(aa_df2) = aa
aa_df3$freq = sum(aa_df3[1,])
aa_df3[[tcrPart]] = tcr.name
rownames(aa_df3) = NULL
colnames(aa_df3)[1:5] = paste0(deparse(substitute(kh)),"_" ,colnames(aa_df3)[1:5])
} else {
aa_df3 = aa_df3$hash.ID %>% table() %>% data.frame() %>% t() %>% data.frame()

colnames(aa_df3)= as.matrix(aa_df3[1,])
theNames = colnames(aa_df3)
aa_df3 = aa_df3[2,] %>% data.frame()
colnames(aa_df3) = theNames


for (i in c("mHTO01","mHTO02","mHTO04","mHTO05")){
  if (!(i %in% colnames(aa_df3))){
    aa_df3[[i]] = 0
  }
}

aa_df3 = apply(aa_df3, 2, as.numeric) %>% data.frame() %>% t() %>% data.frame()
#rownames(aa_df2) = aa
aa_df3$freq = sum(aa_df3[1,])
aa_df3[[tcrPart]] = tcr.name
rownames(aa_df3) = NULL
colnames(aa_df3)[1:5] = paste0(deparse(substitute(kh)),"_" ,colnames(aa_df3)[1:5])
  }
myList[[tcr.name]] = aa_df3
}
my_data = do.call(dplyr::bind_rows, myList)
return(my_data)
}


createFreq = function(tcrPart){
  s1 = myTCR_freqTable(kh = het_1MF1, tcrPart = tcrPart)
  s2 = myTCR_freqTable(kh = het_3MF1, tcrPart = tcrPart)
  s3 = myTCR_freqTable(kh = ko_1MF2, tcrPart = tcrPart)
  s4 = myTCR_freqTable(kh = ko_3MF2, tcrPart = tcrPart)
  s1s2 = merge(s1, s2, tcrPart = tcrPart)
  s3s4 = merge(s3, s4, tcrPart = tcrPart)
  s1s2s3s4 = merge(s1s2, s3s4, tcrPart = tcrPart)
  return(s1s2s3s4)
}



df_cell_TRAV = createFreq("cell_TRAV")
df_cell_TRAJ = createFreq("cell_TRAJ")
df_cell_TRBV = createFreq("cell_TRBV")
df_cell_TRBJ = createFreq("cell_TRBJ")

df_cell_TCRa = createFreq("TCRa")
df_cell_TCRb = createFreq("TCRb")

df_top_alpha = createFreq("cell_top_alpha")
df_top_beta = createFreq("cell_top_beta")
df_ab_pair = createFreq("ab_pair")

df_cell_TCRab = createFreq("TCRab")

write.csv(df_cell_TRAV, "TRAV.csv", quote = F, row.names = F)
write.csv(df_cell_TRAJ, "TRAJ.csv", quote = F, row.names = F)
write.csv(df_cell_TRBV, "TRBV.csv", quote = F, row.names = F)
write.csv(df_cell_TRBJ, "TRBJ.csv", quote = F, row.names = F)
write.csv(df_cell_TCRa, "TCRa.csv", quote = F, row.names = F)
write.csv(df_cell_TCRb, "TCRb.csv", quote = F, row.names = F)
write.csv(df_cell_TCRab, "TCRab.csv", quote = F, row.names = F)
write.csv(df_top_alpha, "alpha_CDR3.csv", quote = F, row.names = F)
write.csv(df_top_beta, "beta_CDR3.csv", quote = F, row.names = F)

```


```{r}
myTCR_heatmap = function(df, sample){
  df[,-1] = apply(df[,-1], 2, function(x){x/sum(x)})
  df_select = select(df, contains(sample))
  rownames(df_select) =df[,1]
  df_select$sum = rowSums(df_select)
  df_select = df_select[which(df_select$sum>0), 1:4]
  
  mat = df_select
  
  heat = 
  ComplexHeatmap::Heatmap(matrix = t(scale(t(mat))), 
                          name="Z-score", km=1, 
                        col=colorRampPalette(rev(brewer.pal(n = 7, name ="RdYlBu")))(100), 
                        row_names_gp = gpar(fontsize = ifelse(nrow(mat) <= 30, 10, 
                                                       ifelse(30<nrow(mat)&nrow(mat) <= 45, 7,
                                                       ifelse(45<nrow(mat)&nrow(mat) <= 60, 5, 4))), fontface = "bold"),
                        column_names_gp = gpar(fontsize = 10, fontface = "bold"),
                        cluster_columns = T, column_title_gp = gpar(fontsize = 13, fontface = "bold"), 
                        column_title_side = "top", column_names_rot = 45, width = unit(5, "cm")) 
  pdf(file = paste0("heatmap_", colnames(df)[1],"_" ,sample, ".pdf"))
  print(heat)
  dev.off()
  print(heat)
  
   # mat = df_select[,2:4]
  mat$name = rownames(mat)
  mat_long = melt(mat)
  
p = 
  ggplot(mat_long, aes(x = variable, y = value, group = 1)) + geom_point() + facet_wrap(vars(name))+ geom_line()+ theme(axis.text.x = element_text(color="#993333",angle=90))
  
pdf(file = paste0("geomLine_",colnames(df)[1],"_" ,sample, ".pdf"), width = 10, height = 11)
print(p)
dev.off()
}

myTCR_heatmap(df = df_cell_TRAV, sample = "mHTO01")
myTCR_heatmap(df = df_cell_TRBV, sample = "mHTO01")
myTCR_heatmap(df = df_cell_TRAJ, sample = "mHTO01")
myTCR_heatmap(df = df_cell_TRBJ, sample = "mHTO01")

myTCR_heatmap(df = df_cell_TRAV, sample = "mHTO02")
myTCR_heatmap(df = df_cell_TRBV, sample = "mHTO02")
myTCR_heatmap(df = df_cell_TRAJ, sample = "mHTO02")
myTCR_heatmap(df = df_cell_TRBJ, sample = "mHTO02")

myTCR_heatmap(df = df_cell_TRAV, sample = "mHTO04")
myTCR_heatmap(df = df_cell_TRBV, sample = "mHTO04")
myTCR_heatmap(df = df_cell_TRAJ, sample = "mHTO04")
myTCR_heatmap(df = df_cell_TRBJ, sample = "mHTO04")

myTCR_heatmap(df = df_cell_TRAV, sample = "mHTO05")
myTCR_heatmap(df = df_cell_TRBV, sample = "mHTO05")
myTCR_heatmap(df = df_cell_TRAJ, sample = "mHTO05")
myTCR_heatmap(df = df_cell_TRBJ, sample = "mHTO05")

myTCR_heatmap(df = df_cell_TRAV, sample = "het_1MF1_mHTO")
myTCR_heatmap(df = df_cell_TRAV, sample = "het_3MF1_mHTO")
myTCR_heatmap(df = df_cell_TRAV, sample = "ko_1MF2_mHTO")
myTCR_heatmap(df = df_cell_TRAV, sample = "ko_3MF2_mHTO")

myTCR_heatmap(df = df_cell_TRAJ, sample = "het_1MF1_mHTO")
myTCR_heatmap(df = df_cell_TRAJ, sample = "het_3MF1_mHTO")
myTCR_heatmap(df = df_cell_TRAJ, sample = "ko_1MF2_mHTO")
myTCR_heatmap(df = df_cell_TRAJ, sample = "ko_3MF2_mHTO")


myTCR_heatmap(df = df_cell_TRBV, sample = "het_1MF1_mHTO")
myTCR_heatmap(df = df_cell_TRBV, sample = "het_3MF1_mHTO")
myTCR_heatmap(df = df_cell_TRBV, sample = "ko_1MF2_mHTO")
myTCR_heatmap(df = df_cell_TRBV, sample = "ko_3MF2_mHTO")

myTCR_heatmap(df = df_cell_TRBJ, sample = "het_1MF1_mHTO")
myTCR_heatmap(df = df_cell_TRBJ, sample = "het_3MF1_mHTO")
myTCR_heatmap(df = df_cell_TRBJ, sample = "ko_1MF2_mHTO")
myTCR_heatmap(df = df_cell_TRBJ, sample = "ko_3MF2_mHTO")


```

```{r}

myTCR_geom_line = function(df, sample){


 df[,-1] = apply(df[,-1], 2, function(x){x/sum(x)})
  df_select = select(df, contains(sample))
  rownames(df_select) =df[,1]
  df_select$sum = rowSums(df_select)
  df_select = df_select[which(df_select$sum>0), 1:4]
  
  mat = df_select[,2:4]
  mat$name = rownames(mat)
  mat_long = melt(mat)
  
p = 
  ggplot(mat_long, aes(x = variable, y = value, group = 1)) + geom_point() + facet_wrap(vars(name))+ geom_line()+ theme(axis.text.x = element_text(color="#993333",angle=90))
  
pdf(file = paste0("geomLine_",colnames(df)[1],"_" ,sample, ".pdf"), width = 9, height = 10)
print(p)
dev.off()

# ggsave(filename = paste0("geomLine_",colnames(df)[1],"_" ,sample, ".pdf"), plot = p)

}


myTCR_geom_line(df = df_cell_TRAV, sample = "het_1MF1_mHTO")


```


```{r}
df_cell_TRAV_perc = df_cell_TRAV
df_cell_TRAV_perc[,-1] = apply(df_cell_TRAV_perc[,-1], 2, function(x){x/sum(x)})

# look at a hashtag across samples, or samples across hashtags
df_cell_TRAV_perc_mHTO02 = select(df_cell_TRAV_perc, contains("mHTO02"))
rownames(df_cell_TRAV_perc_mHTO02) =df_cell_TRAV_perc[,1]
df_cell_TRAV_perc_mHTO02$sum = rowSums(df_cell_TRAV_perc_mHTO02)
df_cell_TRAV_perc_mHTO02 = df_cell_TRAV_perc_mHTO02[which(df_cell_TRAV_perc_mHTO02$sum>0), 1:4]


mat =df_cell_TRAV_perc_mHTO02 

myHeatmap = function(mat){
  

heat = 
ComplexHeatmap::Heatmap(matrix = t(scale(t(mat))), 
                          name="Z-score", km=1, 
                        col=colorRampPalette(rev(brewer.pal(n = 7, name ="RdYlBu")))(100), 
                        row_names_gp = gpar(fontsize = ifelse(nrow(mat) <= 30, 10, 
                                                       ifelse(30<nrow(mat)&nrow(mat) <= 45, 7,
                                                       ifelse(45<nrow(mat)&nrow(mat) <= 60, 5, 4))), fontface = "bold"),
                        column_names_gp = gpar(fontsize = 10, fontface = "bold"),
                        cluster_columns = T, column_title = deparse(substitute(mat)), column_title_gp = gpar(fontsize = 13, fontface = "bold"), 
                        column_title_side = "top", column_names_rot = 45, width = unit(5, "cm")) 
  pdf(file = paste0(deparse(substitute(mat)), ".pdf"))
  print(heat)
  dev.off()
}

myHeatmap(mat = df_cell_TRAV_perc_mHTO02)

```




```{r}
df_ko_TCRa = myTCR_freqTable(kh = ko, tcrPart = "TCRa")
df_het_TCRa = myTCR_freqTable(kh = het, tcrPart = "TCRa")
df_ko.het_TCRa = merge(df_ko_TCRa, df_het_TCRa, by="TCRa")
write.csv(df_ko.het_TCRa, "processedData/TCRa_freq.csv",quote = F, row.names = F)

df_ko_TCRb = myTCR_freqTable(kh = ko, tcrPart = "TCRb")
df_het_TCRb = myTCR_freqTable(kh = het, tcrPart = "TCRb")
df_ko.het_TCRb = merge(df_ko_TCRb, df_het_TCRb, by="TCRb")
write.csv(df_ko.het_TCRb, "processedData/TCRb_freq.csv",quote = F, row.names = F)

df_ko_cell_TRAV = myTCR_freqTable(kh = ko, tcrPart = "cell_TRAV")
df_het_cell_TRAV = myTCR_freqTable(kh = het, tcrPart = "cell_TRAV")
df_ko.het_cell_TRAV = merge(df_ko_cell_TRAV, df_het_cell_TRAV, by="cell_TRAV")
write.csv(df_ko.het_cell_TRAV, "processedData/cell_TRAV_freq.csv",quote = F, row.names = F)

df_ko_cell_TRBV = myTCR_freqTable(kh = ko, tcrPart = "cell_TRBV")
df_het_cell_TRBV = myTCR_freqTable(kh = het, tcrPart = "cell_TRBV")
df_ko.het_cell_TRBV = merge(df_ko_cell_TRBV, df_het_cell_TRBV, by="cell_TRBV")
write.csv(df_ko.het_cell_TRBV, "processedData/cell_TRBV_freq.csv",quote = F, row.names = F)

df_ko_cell_TRAJ = myTCR_freqTable(kh = ko, tcrPart = "cell_TRAJ")
df_het_cell_TRAJ = myTCR_freqTable(kh = het, tcrPart = "cell_TRAJ")
df_ko.het_cell_TRAJ = merge(df_ko_cell_TRAJ, df_het_cell_TRAJ, by="cell_TRAJ")
write.csv(df_ko.het_cell_TRAJ, "processedData/cell_TRAJ_freq.csv",quote = F, row.names = F)

df_ko_cell_TRBJ = myTCR_freqTable(kh = ko, tcrPart = "cell_TRBJ")
df_het_cell_TRBJ = myTCR_freqTable(kh = het, tcrPart = "cell_TRBJ")
df_ko.het_cell_TRBJ = merge(df_ko_cell_TRBJ, df_het_cell_TRBJ, by="cell_TRBJ")
write.csv(df_ko.het_cell_TRBJ, "processedData/cell_TRBJ_freq.csv",quote = F, row.names = F)

df_ko_cell_top_alpha = myTCR_freqTable(kh = ko, tcrPart = "cell_top_alpha")
df_het_cell_top_alpha = myTCR_freqTable(kh = het, tcrPart = "cell_top_alpha")
df_ko.het_cell_top_alpha = merge(df_ko_cell_top_alpha, df_het_cell_top_alpha, by="cell_top_alpha")
write.csv(df_ko.het_cell_top_alpha, "processedData/cell_top_alpha_freq.csv",quote = F, row.names = F)

df_ko_cell_top_beta = myTCR_freqTable(kh = ko, tcrPart = "cell_top_beta")
df_het_cell_top_beta = myTCR_freqTable(kh = het, tcrPart = "cell_top_beta")
df_ko.het_cell_top_beta = merge(df_ko_cell_top_beta, df_het_cell_top_beta, by="cell_top_beta")
write.csv(df_ko.het_cell_top_beta, "processedData/cell_top_beta_freq.csv",quote = F, row.names = F)

df_ko_ab_pair = myTCR_freqTable(kh = ko, tcrPart = "ab_pair")
df_het_ab_pair = myTCR_freqTable(kh = het, tcrPart = "ab_pair")
df_ko.het_ab_pair = merge(df_ko_ab_pair, df_het_ab_pair, by="ab_pair")
write.csv(df_ko.het_ab_pair, "processedData/ab_pair_freq.csv",quote = F, row.names = F)
```




```{r}
hetf = ko_het_alpha_freq[which(ko_het_alpha_freq$ko_freq==0),]
hetf = ko_het_beta_freq[which(ko_het_beta_freq$ko_freq==0),]

hetf_select = hetf[which(hetf$het_freq>1),1]
so@meta.data$het.unique = ifelse(so@meta.data[[tcr]]%in%hetf_select, "yes","no")
hetf = melt(hetf[1:20,])

kof = ko_het_alpha_freq[which(ko_het_alpha_freq$het_freq==0),]
kof = ko_het_beta_freq[which(ko_het_beta_freq$het_freq==0),]

kof_select = kof[which(kof$ko_freq>1),1]
so@meta.data$ko.unique = ifelse(so@meta.data[[tcr]]%in%kof_select, "yes","no")
kof = melt(kof[1:20,])



ggplot(hetf, aes(x = Var1, y = value, fill=variable)) + geom_bar(stat = "identity", position = "dodge") + coord_flip()+
  ylab("Frequency") + xlab(tcr) + theme(legend.position="top")
DimPlot(so, group.by = "het.unique", repel = TRUE, reduction = "tsne",cols = c("lightgrey", "red"))+ 
ggtitle(paste("selected",tcr, "in b5t+/-"))+ theme(legend.position="top")


```


```{r}
alpha_select = subset(so, cells=which(so@meta.data$het.unique=="yes"|so@meta.data$ko.unique=="yes"))

DimPlot(alpha_select,group.by = "ko.unique", reduction = "tsne")
alpha_select@meta.data
```


```{r}
Idents(alpha_select) = "ko.unique"
ko.unique.markers =
FindMarkers(alpha_select, ident.1="yes", ident.2="no",test.use="MAST", logfc.threshold = 0.25, min.pct = 0.25)

ko.unique.markers.heatmap = DoHeatmap(alpha_select, features = rownames(ko.unique.markers))+ NoLegend()

ko.unique.aveExpr=
AverageExpression(alpha_select, features = rownames(ko.unique.markers))

mat =ko.unique.aveExpr$RNA



ComplexHeatmap::Heatmap(matrix = t(scale(t(mat))), 
                          name="Z-score", km=1, 
                        col=colorRampPalette(rev(brewer.pal(n = 7, name ="RdYlBu")))(100), 
                        row_names_gp = gpar(fontsize = ifelse(nrow(mat) <= 30, 10, 
                                                       ifelse(30<nrow(mat)&nrow(mat) <= 45, 7,
                                                       ifelse(45<nrow(mat)&nrow(mat) <= 60, 5, 4))), fontface = "bold"),
                        column_names_gp = gpar(fontsize = 10, fontface = "bold"),
                        cluster_columns = T, column_title = "gene expression", column_title_gp = gpar(fontsize = 13, fontface = "bold"), 
                        column_title_side = "top", column_names_rot = 45, width = unit(5, "cm")) 

so@meta.data

```


```{r}
het_HTO01 = het[which(het$HTO_maxID=="mHTO01"&het[[tcr]]!="NA&NA&NA"),]
het_HTO02 = het[which(het$HTO_maxID=="mHTO02"&het[[tcr]]!="NA&NA&NA"),]
het_HTO04 = het[which(het$HTO_maxID=="mHTO04"&het[[tcr]]!="NA&NA&NA"),]
het_HTO05 = het[which(het$HTO_maxID=="mHTO05"&het[[tcr]]!="NA&NA&NA"),]


ko_HTO01 = ko[which(ko$HTO_maxID=="mHTO01"&ko[[tcr]]!="NA&NA&NA"),]
ko_HTO02 = ko[which(ko$HTO_maxID=="mHTO02"&ko[[tcr]]!="NA&NA&NA"),]
ko_HTO04 = ko[which(ko$HTO_maxID=="mHTO04"&ko[[tcr]]!="NA&NA&NA"),]
ko_HTO05 = ko[which(ko$HTO_maxID=="mHTO05"&ko[[tcr]]!="NA&NA&NA"),]

write.csv(het_HTO05, "het_HTO05.csv", quote = F)

so@meta.data

Idents(so) = "HTO_maxID"
DimPlot(so, reduction = "tsne")

het_HTO01[[tcr]] %>% table() %>%plot(las=2)
het_HTO02[[tcr]] %>% table() %>%plot(las=2)
het_HTO04[[tcr]] %>% table() %>%plot(las=2)
het_HTO05[[tcr]] %>% table() %>%plot(las=2)
library(VennDiagram)
venn.plot <- venn.diagram(list(het_HTO01[[tcr]],
                                     het_HTO02[[tcr]],
                                     het_HTO04[[tcr]],
                                     het_HTO05[[tcr]]), NULL, 
                                fill=c("blue", "red","grey","green"), lwd = 1, cat.cex = 2,
                                alpha=c(0.5,0.5,0.5,0.5), cex = 3, cat.fontface=4, 
                                category.names=c("het_HTO01","het_HTO02","het_HTO04","het_HTO05"))
#Het
venn.plot <- venn.diagram(list(het_HTO02[[tcr]],
                                     het_HTO04[[tcr]],
                                     het_HTO05[[tcr]]), NULL, 
                                fill=c("blue", "red","grey"), lwd = 1, cat.cex = 2,
                                alpha=c(0.5,0.5,0.5), cex = 3, cat.fontface=4, 
                                category.names=c("het_HTO02","het_HTO04","het_HTO05"))

#ko
venn.plot <- venn.diagram(list(ko_HTO02[[tcr]],
                                     ko_HTO04[[tcr]],
                                     ko_HTO05[[tcr]]), NULL, 
                                fill=c("blue", "red","grey"), lwd = 1, cat.cex = 2,
                                alpha=c(0.5,0.5,0.5), cex = 3, cat.fontface=4, 
                                category.names=c("ko_HTO02","ko_HTO04","ko_HTO05"))

plot_grid(venn.plot)
```


```{r}
het = so@meta.data[which(so@meta.data$sample_name%in%c("1MF1","3MF1")),]
het = het[-which(is.na(het$cell_TRAV)),]
ko = so@meta.data[which(so@meta.data$sample_name%in%c("1MF2","3MF2")),]
ko = ko[-which(is.na(ko$cell_TRAV)),]

het_1MF1 = so@meta.data[which(so@meta.data$sample_name%in%c("1MF1")),]
het_3MF1 = so@meta.data[which(so@meta.data$sample_name%in%c("3MF1")),]
ko_1MF2 = so@meta.data[which(so@meta.data$sample_name%in%c("1MF2")),]
ko_3MF2 = so@meta.data[which(so@meta.data$sample_name%in%c("3MF2")),]

het_1MF1 = het[which(het$sample_name%in%c("1MF1")),]
het_3MF1 = het[which(het$sample_name%in%c("3MF1")),]
ko_1MF2 =  ko[which(ko$sample_name%in%c("1MF2")),]
ko_3MF2 =  ko[which(ko$sample_name%in%c("3MF2")),]

myFrequentTCR = function(tcr,n){
  all = c(het_1MF1[[tcr]],het_3MF1[[tcr]],ko_1MF2[[tcr]],ko_3MF2[[tcr]]) %>% table()%>%data.frame()
  all = all[order(all$Freq, decreasing = T),]
  colnames(all) = paste(tcr,colnames(all), sep = "_")
  colnames(all)[1] = str_sub(string = colnames(all)[1],start = 1,end = -3)
  all_top = all[1:n,1]
  return(all_top)
}

all_TRAV_Top = myFrequentTCR(tcr = "cell_TRAV", n = 40)
all_TRBV_Top = myFrequentTCR(tcr = "cell_TRBV", n = 22)
all_alpha_CDR3_Top = myFrequentTCR(tcr = "cell_top_alpha", n = 40) #CDR3 sequence


tcr = "cell_TRBV"
all = c(het_1MF1[[tcr]],het_3MF1[[tcr]],ko_1MF2[[tcr]],ko_3MF2[[tcr]]) %>% table()%>%data.frame()
```


```{r}
all_TRAV = c(het_1MF1$cell_TRAV,het_3MF1$cell_TRAV,ko_1MF2$cell_TRAV,ko_3MF2$cell_TRAV) %>% table()%>%data.frame()
all_TRAV = all_TRAV[order(all_TRAV$Freq, decreasing = T),]
colnames(all_TRAV) = paste("all_TRAV",colnames(all_TRAV), sep = "_")
colnames(all_TRAV)[1] = str_sub(string = colnames(all_TRAV)[1],start = 1,end = -3)
all_TRAV_Top20 = all_TRAV[1:40,1]

myTCR = function(tcr,sample, top.part){
  sample.part = sample[[tcr]] %>% table()%>%data.frame() 
  sample.part$fraction = sample.part[,2]/length(sample[[tcr]])
  colnames(sample.part) = paste(deparse(substitute(sample)),colnames(sample.part), sep = "_")
  colnames(sample.part)[1] = str_sub(string = colnames(sample.part)[1],start = 1,end = -3)
  sample.part_n = sample.part[which(sample.part[,1]%in%top.part),c(1,3)]
  colnames(sample.part_n)[1] =  tcr
  return(sample.part_n)
}

#TRAV
het_1MF1_TRAV_20 = myTCR(tcr = "cell_TRAV",sample = het_1MF1,top.part = all_TRAV_Top20)
het_3MF1_TRAV_20 = myTCR(tcr = "cell_TRAV",sample = het_3MF1,top.part = all_TRAV_Top20)
ko_1MF2_TRAV_20 = myTCR(tcr = "cell_TRAV",sample = ko_1MF2,top.part = all_TRAV_Top20)
ko_3MF2_TRAV_20 = myTCR(tcr = "cell_TRAV",sample = ko_3MF2,top.part = all_TRAV_Top20)

TRAV = merge(het_1MF1_TRAV_20, het_3MF1_TRAV_20 ,by="cell_TRAV")%>%
merge(ko_1MF2_TRAV_20,by="cell_TRAV") %>%
merge(ko_3MF2_TRAV_20 ,by="cell_TRAV")
colnames(TRAV) = gsub("_fraction","",colnames(TRAV))
rownames(TRAV) = TRAV$cell_TRAV
TRAV$cell_TRAV = NULL

#TRBV
het_1MF1_TRBV_20 = myTCR(tcr = "cell_TRBV",sample = het_1MF1,top.part = all_TRBV_Top)
het_3MF1_TRBV_20 = myTCR(tcr = "cell_TRBV",sample = het_3MF1,top.part = all_TRBV_Top)
ko_1MF2_TRBV_20 = myTCR(tcr = "cell_TRBV",sample = ko_1MF2,top.part = all_TRBV_Top)
ko_3MF2_TRBV_20 = myTCR(tcr = "cell_TRBV",sample = ko_3MF2,top.part = all_TRBV_Top)

TRBV = merge(het_1MF1_TRBV_20, het_3MF1_TRBV_20 ,by="cell_TRBV")%>%
merge(ko_1MF2_TRBV_20,by="cell_TRBV") %>%
merge(ko_3MF2_TRBV_20 ,by="cell_TRBV")
colnames(TRBV) = gsub("_fraction","",colnames(TRBV))
rownames(TRBV) = TRBV$cell_TRBV
TRBV$cell_TRBV = NULL


#alpha CDR3
het_1MF1_alpha_CDR3_Top = myTCR(tcr = "cell_top_alpha",sample = het_1MF1,top.part = all_alpha_CDR3_Top)
het_3MF1_alpha_CDR3_Top = myTCR(tcr = "cell_top_alpha",sample = het_3MF1,top.part = all_alpha_CDR3_Top)
ko_1MF2_alpha_CDR3_Top = myTCR(tcr = "cell_top_alpha",sample = ko_1MF2,top.part = all_alpha_CDR3_Top)
ko_3MF2_alpha_CDR3_Top = myTCR(tcr = "cell_top_alpha",sample = ko_3MF2,top.part = all_alpha_CDR3_Top)



alpha_CDR3 = merge(het_1MF1_alpha_CDR3_Top, het_3MF1_alpha_CDR3_Top ,by="cell_top_alpha")%>%
merge(ko_1MF2_alpha_CDR3_Top,by="cell_top_alpha") %>%
merge(ko_3MF2_alpha_CDR3_Top ,by="cell_top_alpha")
colnames(alpha_CDR3) = gsub("_fraction","",colnames(alpha_CDR3))
rownames(alpha_CDR3) = alpha_CDR3$cell_top_alpha
alpha_CDR3$cell_top_alpha = NULL



TRAV_long = melt(TRAV)
ggplot(TRAV_long, aes(x = cell_TRAV, y = value, fill=variable)) + geom_bar(stat = "identity", position = "dodge") + coord_flip()



tcr = "cell_TRAV"
sample = het_1MF1


sample.part = sample[[tcr]] %>% table()%>%data.frame() 
sample.part$fraction = sample.part[,2]/length(sample[[tcr]])


colnames(het_1MF1_TRAV) = paste("het_1MF1_TRAV",colnames(het_1MF1_TRAV), sep = "_")
colnames(het_1MF1_TRAV)[1] = str_sub(string = colnames(het_1MF1_TRAV)[1],start = 1,end = -3)
het_1MF1_TRAV_20 = het_1MF1_TRAV[which(het_1MF1_TRAV[,1]%in%all_TRAV_Top20),c(1,3)]
colnames(het_1MF1_TRAV_20)[1] = "cell_TRAV"


het_3MF1_TRAV = het_3MF1$cell_TRAV %>% table()%>%data.frame()
ko_1MF2_TRAV = ko_1MF2$cell_TRAV %>% table()%>%data.frame()
ko_3MF2_TRAV = ko_3MF2$cell_TRAV %>% table()%>%data.frame()


so@meta.data

```
